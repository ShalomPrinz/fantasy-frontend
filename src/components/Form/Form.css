.form-text-input {
  padding: 10px;
  border: 2px solid black;
  background-color: white;
  color: #2d3748;
  border-radius: 10px;
}

.form-text-input:invalid {
  border: 4px solid #ff7d87;
  box-shadow: none;
}

.form-text-input:focus {
  outline: none;
}
