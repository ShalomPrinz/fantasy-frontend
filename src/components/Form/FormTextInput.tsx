import { useField } from "formik";

export type TextInput = {
  label: string;
  name: string;
  placeholder?: string;
  type: string;
};

interface FormTextInputProps {
  name: string;
  type: string;
  [x: string]: string;
}

export const FormTextInput = ({ label, ...props }: FormTextInputProps) => {
  const [field, meta] = useField(props);
  if (props.type === "password") props["autoComplete"] = "on";
  return (
    <>
      <label htmlFor={props.name}>{label}</label>
      <input className="form-text-input" {...field} {...props} />
      {meta.touched && meta.error ? (
        <div className="text-danger">{meta.error}</div>
      ) : null}
    </>
  );
};
