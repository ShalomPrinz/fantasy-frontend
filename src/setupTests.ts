// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import axios from "axios";
import { setupServer } from "msw/node";

import { handlers } from "../test";
import * as TeamContext from "./contexts/TeamContext";
import * as UserContext from "./features/authentication/contexts/UserContext";
import { LeagueInfo, Message, Player, Team, User } from "./types";

jest.setTimeout(10000);

axios.defaults.baseURL = "";

const server = setupServer(...handlers);
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

interface MockedUserContext {
  logout?: () => Promise<void>;
  state?: UserContext.UserState;
  user?: User;
}

interface MockedUser {
  inbox?: Message[];
  leagues?: LeagueInfo[];
  name?: string;
  players?: Player[];
}

export const getTestUser = (user?: MockedUser): User => ({
  id: "id",
  inbox: user?.inbox || [],
  leagues: user?.leagues || [],
  name: user?.name || "",
  team: new Team(user?.players || []),
});

export const mockUser = (context: MockedUserContext) => {
  const userContext = jest.spyOn(UserContext, "useUser");

  const { logout, state, user } = context;
  const mockedLogout = typeof logout !== "undefined" ? logout : async () => {};
  const mockedState =
    typeof state !== "undefined" ? state : UserContext.UserState.LOGGED_USER;
  const mockedUser = typeof user !== "undefined" ? user : getTestUser();

  userContext.mockImplementation(() => ({
    loading: false,
    login: async () => {},
    logout: mockedLogout,
    register: async () => {},
    state: mockedState,
    user: mockedUser,
    invalidate: async () => {},
  }));

  return userContext.mockRestore;
};

export const mockTeam = (team: Team) => {
  const teamContext = jest.spyOn(TeamContext, "useTeamState");
  teamContext.mockImplementation(() => team);
};

export * from "../test";
