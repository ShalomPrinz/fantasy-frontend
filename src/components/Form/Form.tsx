import { Form, Formik } from "formik";
import { Row } from "react-bootstrap";
import { SchemaLike } from "yup/lib/types";

import { ConditionalList } from "../";
import type { TextInput } from "./FormTextInput";
import { FormTextInput } from "./FormTextInput";
import "./Form.css";

interface FormProps {
  onSubmit: (values: {}) => void;
  schema: SchemaLike;
  submitDisabled: boolean;
  textInputs: Array<TextInput>;
  title: string;
}

const AppForm = ({
  onSubmit,
  schema,
  submitDisabled,
  textInputs,
  title,
}: FormProps) => {
  const initialValues = textInputs.reduce(
    (o, key) => ({ ...o, [key.name]: "" }),
    {}
  );

  const spaceBetweenInputs = textInputs.length === 1 ? "4" : "2";
  const inputCallback = (item: TextInput) => (
    <Row className={`mx-2 px-2 py-${spaceBetweenInputs}`}>
      <FormTextInput {...item} />
    </Row>
  );

  return (
    <Formik
      initialValues={initialValues}
      validationSchema={schema}
      onSubmit={(values, { setSubmitting }) => {
        onSubmit(values);
        setSubmitting(false);
      }}
    >
      <Form className="container">
        <Row>
          <h1>{title}</h1>
        </Row>
        <ConditionalList itemCallback={inputCallback} list={textInputs} />
        <Row>
          <button
            className={`fs-2 bg-default py-2 px-auto mt-3 rounded button-border-focus${
              submitDisabled ? " text-muted" : ""
            }`}
            disabled={submitDisabled}
            type="submit"
          >
            Submit
          </button>
        </Row>
      </Form>
    </Formik>
  );
};

export default AppForm;
