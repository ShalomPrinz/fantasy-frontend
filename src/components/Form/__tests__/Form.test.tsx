import { act, clickElement, render, screen, typeElement } from "setupTests";
import * as Yup from "yup";

import Form from "../Form";

const textInput = [
  {
    id: 0,
    label: "first input",
    name: "firstInput",
    type: "text",
  },
];

describe("Form", () => {
  it("should render Form component", () => {
    const { asFragment } = render(
      <Form
        onSubmit={() => {}}
        schema={Yup.object()}
        submitDisabled={false}
        textInputs={[]}
        title={"Test Form"}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render Form component with text input", () => {
    const { asFragment } = render(
      <Form
        onSubmit={() => {}}
        schema={Yup.object()}
        submitDisabled={false}
        textInputs={textInput}
        title={"Test Form"}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe("when user inserts input", () => {
    it("should validate values and report errors according to the given schema", async () => {
      const minChars = 5;
      const errorMessage = `Must be at least ${minChars} characters`;

      const { user } = render(
        <Form
          onSubmit={() => {}}
          schema={Yup.object({
            firstInput: Yup.string().min(minChars, errorMessage),
          })}
          submitDisabled={false}
          textInputs={textInput}
          title={"Test Form"}
        />
      );

      let errorElement = screen.queryByText(errorMessage);
      expect(errorElement).toBe(null);

      const input = screen.getByRole("textbox");
      await typeElement(user, input, "a".repeat(minChars - 1));
      act(() => input.blur());

      errorElement = screen.queryByText(errorMessage);
      expect(errorElement).toBeInTheDocument();

      await typeElement(user, input, "a".repeat(minChars));

      errorElement = screen.queryByText(errorMessage);
      expect(errorElement).toBe(null);
    });
  });

  describe("when user submits form", () => {
    it("should call onSubmit with the user inserted values", async () => {
      let value = {
        firstInput: "",
      };

      const { user } = render(
        <Form
          onSubmit={(v) => {
            // @ts-ignore
            value = v;
          }}
          schema={Yup.object()}
          submitDisabled={false}
          textInputs={textInput}
          title={"Test Form"}
        />
      );

      const inputString = "SOME INPUT";

      const input = screen.getByRole("textbox");
      await typeElement(user, input, inputString);

      expect(value.firstInput).toEqual("");

      const submitButton = screen.getByRole("button");
      await clickElement(user, submitButton);

      expect(value.firstInput).toEqual(inputString);
    });

    it("should not allow submit if submit is disabled", async () => {
      let value = {
        firstInput: "",
      };

      const { user } = render(
        <Form
          onSubmit={(v) => {
            // @ts-ignore
            value = v;
          }}
          schema={Yup.object()}
          submitDisabled={true}
          textInputs={textInput}
          title={"Test Form"}
        />
      );

      const inputString = "SOME INPUT";

      const input = screen.getByRole("textbox");
      await typeElement(user, input, inputString);

      expect(value.firstInput).toEqual("");

      const submitButton = screen.getByRole("button");
      await clickElement(user, submitButton);

      expect(value.firstInput).toEqual("");
    });
  });
});
