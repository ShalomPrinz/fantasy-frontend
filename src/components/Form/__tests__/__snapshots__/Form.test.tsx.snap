// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render Form component 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="container"
  >
    <div
      class="row"
    >
      <h1>
        Test Form
      </h1>
    </div>
    <div
      class="row"
    >
      <button
        class="fs-2 bg-default py-2 px-auto mt-3 rounded button-border-focus"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`Form should render Form component with more than one input 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="container"
  >
    <div
      class="row"
    >
      <h1>
        Test Form
      </h1>
    </div>
    <div
      class="mx-2 px-2 py-2 row"
    >
      <label
        for="firstInput"
      >
        first input
      </label>
      <input
        class="form-text-input"
        id="0"
        name="firstInput"
        type="text"
        value=""
      />
    </div>
    <div
      class="mx-2 px-2 py-2 row"
    >
      <label
        for="secondInput"
      >
        second input
      </label>
      <input
        class="form-text-input"
        id="1"
        name="secondInput"
        type="text"
        value=""
      />
    </div>
    <div
      class="row"
    >
      <button
        class="fs-2 bg-default py-2 px-auto mt-3 rounded button-border-focus"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`Form should render Form component with text input 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="container"
  >
    <div
      class="row"
    >
      <h1>
        Test Form
      </h1>
    </div>
    <div
      class="mx-2 px-2 py-4 row"
    >
      <label
        for="firstInput"
      >
        first input
      </label>
      <input
        class="form-text-input"
        id="0"
        name="firstInput"
        type="text"
        value=""
      />
    </div>
    <div
      class="row"
    >
      <button
        class="fs-2 bg-default py-2 px-auto mt-3 rounded button-border-focus"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</DocumentFragment>
`;
